addons/binding/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/handler/XiaomiBridgeHandler.java

```diff
@@ -0,0 +1,356 @@
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+/**
+ * The {@link XiaomiBridgeHandler} is responsible for handling commands, which are
+ * sent to one of the channels for the bridge.
+ *
+ * @author Patrick Boos - Initial contribution
+ * @author Dimalo - added device update from heartbeat
+ */
+public class XiaomiBridgeHandler extends ConfigStatusBridgeHandler implements XiaomiSocketListener {
+
+    private static final int DISCOVERY_LOCK_TIME = 10000;
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES = Collections.singleton(THING_TYPE_BRIDGE);
+    private static final JsonParser PARSER = new JsonParser();
+
+    private Logger logger = LoggerFactory.getLogger(this.getClass().getName());
+
+    private List<XiaomiItemUpdateListener> itemListeners = new ArrayList<>();
+    private List<XiaomiItemUpdateListener> itemDiscoveryListeners = new ArrayList<>();
+
+    private String token; // token of gateway
+    private long lastDiscoveryTime;
+    private Map<String, Long> lastOnlineMap = new HashMap<>();
+
+    private Configuration config;
+    private InetAddress host;
+    private int port;
+    private XiaomiBridgeSocket socket;
+
+    public XiaomiBridgeHandler(Bridge bridge) {
+        super(bridge);
+    }
+
+    @Override
+    public Collection<ConfigStatusMessage> getConfigStatus() {
+        // Currently we have no errors. Since we always use discover, it should always be okay.
+        return Collections.emptyList();
+    }
+
+    @Override
+    public void initialize() {
+        try {
+            config = getThing().getConfiguration();
+            host = InetAddress.getByName(config.get(HOST).toString());
+            port = getConfigInteger(config, PORT);
+        } catch (UnknownHostException e) {
+            logger.error("Bridge IP/PORT config is not set or not valid");
+        }
+        /*
+         * TODO:make this code work, for now deactivated - seems it's confusing itself when restarting the binding:
+         * The Handler takes a socket, which is then closed right after that
+         *
+         * // Use existing socket for this port (if one exists)
+         * ArrayList<XiaomiSocket> sockets = XiaomiSocket.getOpenSockets();
+         * logger.debug("Open Sockets are {}", sockets.toString());
+         * if (sockets != null && !(sockets.isEmpty())) {
+         * for (XiaomiSocket s : sockets) {
+         * logger.debug("Checking existing socket this BridgeHandler");
+         * if (s.getPort() == port) {
+         * logger.debug("Using existing socket on port {} for this BridgeHandler", port);
+         * socket = (XiaomiBridgeSocket) s;
+         * break;
+         * }
+         * socket = new XiaomiBridgeSocket(port);
+         * }
+         * } else {
+         * socket = new XiaomiBridgeSocket(port);
+         * }
+         */
+
+        socket = new XiaomiBridgeSocket(port);
```

Comment:
yes, in the dispose methode with unregisterListener. See Javadoc for that one.
✅